### PR TITLE
test(backup): reusable testkit that verifies backup store behavior

### LINKED ISSUE #10254: `AbstractBackupStoreIT` tests are detected as duplicate test runs
On Jenkins we run a script to detect duplicate test runs. The `S3BackupStoreIT` tests are abstract and run for two S3 implementations. This is then detected as duplicate runs which makes the Jenkins pipeline fail.

Open questions:
* Do we want to use the same script on GHA?
* Does failsafe handle this correctly so that one test run doesn't overwrite the other?

### PR DESCRIPTION
**Pull request by lenaschoenburg**
## Description

* Adds a new module with a high level [test interface](https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-interfaces-and-default-methods) `BackupStoreTestKit` that contains tests that verify behavior that should apply to all backup store implementations. These tests are based on the s3 backup store tests.
* Uses this testkit for the s3 backup store implementation.

Some bare bones documentation can be found in the [README](https://github.com/camunda/zeebe/blob/os-backup-store-testkit/backup-stores/testkit/README.md).

## Related issues

closes #10254 because the MinIO and Localstack integration tests are no longer extending an abstract test class with nested test classes.

### CODE REVIEW

**Comment by oleschoenburg** [comment]
@megglos As discussed, feel free to take over here and add any review hints yourself. If you see any larger issues we can also keep this PR open until I'm back and try and find a fix for #10254 in the meantime.

File: backup-stores/testkit/README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  @Override
+  public BackupStore getStore() {
+    return store;
+  }
+}
+```
+
+The testkit requires a clean state for each test, so let's make sure that we use a new GCS bucket
+every for every test:
```
Comment (megglos) [inline backup-stores/testkit/README.md:66]: 🔧 
```suggestion
The testkit requires a clean state for each test, so let's make sure that we use a new GCS bucket
for every test:
```

File: backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/DeletingBackup.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import io.camunda.zeebe.backup.common.BackupIdentifierImpl;
+import io.camunda.zeebe.backup.testkit.support.TestBackupProvider;
+import java.time.Duration;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+public interface DeletingBackup {
+  BackupStore getStore();
```
Comment (megglos) [inline backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/DeletingBackup.java:23]: 💭 NABD could be extracted to another interface `BackupStoreTestBase` to maintain this in one place? Maybe not worth it now though

File: backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/SavingBackup.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  @ArgumentsSource(TestBackupProvider.class)
+  default void backupFailsIfBackupAlreadyExists(final Backup backup) {
+    // when
+    getStore().save(backup).join();
+
+    // then
+    Assertions.assertThat(getStore().save(backup)).failsWithin(Duration.ofSeconds(10));
+    // TODO
+    //  .withThrowableOfType(Throwable.class);
+    //  .withRootCauseInstanceOf(BackupInInvalidStateException.class);
```
Comment (megglos) [inline backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/SavingBackup.java:40]: ❓ guess the TODO is here as this depends on a specific class from the s3 module?

would add an interface method to obtain the expected class here, wdyt?
`Class<? extends Exception> getBackupInInvalidStateExceptionClass();`

resulting in 
```
        .withThrowableOfType(Throwable.class)
        .withRootCauseInstanceOf(getBackupInInvalidStateExceptionClass());
```

File: backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/DeletingBackup.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import io.camunda.zeebe.backup.api.BackupStore;
+import io.camunda.zeebe.backup.common.BackupIdentifierImpl;
+import io.camunda.zeebe.backup.testkit.support.TestBackupProvider;
+import java.time.Duration;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+public interface DeletingBackup {
```
Comment (megglos) [inline backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/DeletingBackup.java:22]: 💭 NABD wdyt about adding a suffix like `TestCases` to make the purpose clear?

**Review by megglos — APPROVED** [review]
LGTM! I actually like the solution with the interfaces more than the nested classes as the scenarios are easier composable that way and that there is no single uber class with lots of cases in it.

One thing I'm not fully sure about is the naming in that case but something that can be followed up on.

I stumbled over one TODO for which I will add a commit to the PR before merging

File: backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/SavingBackup.java (hunk 3 above)
Comment (megglos) [reply backup-stores/testkit/src/main/java/io/camunda/zeebe/backup/testkit/SavingBackup.java]: addressed with https://github.com/camunda/zeebe/pull/10269/commits/fb64966633d2245f24913036631773dc2a5b97f7

**Comment by deepthidevaki** [comment]
@megglos  You are added as reviewer. Please try again.

**Pull request merged**